Adding scripts for ATLAS HEPMC Open Data handling

zlmarshall opened the pull request:
This is the first iteration on scripts required for handling of the new ATLAS event generation open data in HEPMC format. The README explains all the different scripts and files included here (or should).

A few samples have been transferred to CERN already to establish the functionality of all the scripts. Everything seems to be ok so far.

The key outstanding item (needed before anything can actually go onto the Open Data Portal) is the record ID and DOI list for all the various records that will be created. Otherwise this should be just about ready to go, at least to the QA portal for checking.

Commit:
Adding scripts for ATLAS HEPMC Open Data handling

This is the first iteration on scripts required for handling of the new
ATLAS event generation open data in HEPMC format. The README explains
all the different scripts and files included here (or should).

A few samples have been transferred to CERN already to establish the
functionality of all the scripts. Everything seems to be ok so far.

The key outstanding item (needed before anything can actually go onto
the Open Data Portal) is the record ID and DOI list for all the various
records that will be created. Otherwise this should be just about ready
to go, at least to the QA portal for checking.

Commit:
Preferring MC16 to MC15 metadata

Connected to ATLAS internal discussion in
https://its.cern.ch/jira/browse/CENTRPAGE-569

For now, I think this is doing the correct thing

Commit:
Fixing production spreadsheet header after csv move

Commit:
Adding a few features and samples

- Adding late-requested exotics datasets as an explicit list of
  datasets, so the various parsing scripts had to be updated accordingly
- Updating production sheet and metadata requests accordingly
- Sorting keywords in metadata (looks nicer)
- Updating sample rules to appropriately handle the new exotics samples

Commit:
Adding script for building js file from metadata csv

This js file then goes into the ATLAS open data website to document the
metadata. Including it here so that everything is in one place. Because
it is really just a copy of the csv with a header and footer, not trying
to include the output here as well (just unnecessary extra files)

Commit:
Final updates for first release of a test page

This sets the remaining infrastructure up. I believe it is sufficient
for the first release of a test record page. Once more production is
done, we can release more of the records.

zlmarshall commented:
Hi @tiborsimko ,

Thank you for the DOIs and record IDs in
https://github.com/cernopendata/opendata.cern.ch/pull/3737

I've updated the script here to make use of them, and tried to build a little infrastructure so that we don't screw up the assignment or get things mixed up in the future. I've tested a bit, and things seem to be working so far.

I think this means we're ready for a test deployment on the QA instance. This is going to feel like a LOT of infrastructure for just two records, but I am hoping that all this setup means we will be able to scale from 2 to 200 without any problems.

The nominal plan (I hope you agree!) would be to get this up on the QA instance, check things over, then to actually release it onto the normal portal. We'd then let phenomenologists look at it and see how to use it, and assuming everything is kosher we would then go ahead and release the next (larger) batch.

Thanks again,
Zach

tiborsimko commented on atlas-2025-odfr-hepmc/create_file_metadata.py:
```diff
@@ -0,0 +1,91 @@
+        # Update the map of datasets : files
+        datasets[dataset] += [ a['name'] ]
+        # Get the first part of the per-file metadata
+        file_locations[dataset][ a['scope']+':'+a['name'] ] = { 'checksum':'adler32:'+a['adler32'], 'size':a['bytes'], 'events':a['events'], 'type':'HEPMC' }
```
The generated snippets look like:

```console
$ head -17 test/MC_301000_hepmc_filelist.json
[
  {
    "checksum": "adler32:ab2d6244",
    "events": 10000,
    "filename": "mc16_13TeV:HEPMC.43646131._000001.tar.gz.1",
    "size": 519218206,
    "type": "HEPMC",
    "uri_root": "root://eospublic.cern.ch//eos/opendata/atlas/rucio/mc16_13TeV/HEPMC.43646131._000001.tar.gz.1"
  },
  {
    "checksum": "adler32:061ba84a",
    "events": 10000,
    "filename": "mc16_13TeV:HEPMC.43646131._000002.tar.gz.1",
    "size": 517825351,
    "type": "HEPMC",
    "uri_root": "root://eospublic.cern.ch//eos/opendata/atlas/rucio/mc16_13TeV/HEPMC.43646131._000002.tar.gz.1"
  },
```

Some questions and notes:

- Do you want to use `somefile.tar.gz.1` extension?  I guess `.hepmc` might be more welcoming to casual visitors for quick peruse or for remote access? (At the price of consuming more disk space.)
- Note that the generated structure differs a bit from what we are using on the portal, e.g. there is no `events`, e.g. `uri_root` is just `uri`. But I guess we'll create file indexes for these data, as in the past, so the file won't be used directly "as is", but via created file indexes (see next point). If so, then there is no need to change it.
- It seems that all the HEPMC files will be living in the same directory. Is that right also for the records to come? If so, I could create file indexes around `HEPMC.<number>*`, i.e. all `HEPMC.43646541.*` files will be grouped together, all `HEPMC.43646541.*` files together, etc. Looking at the generated files `MC_301000_hepmc_filelist.json`, `MC_301001_hepmc_filelist.json`, etc, this seems to be the case. Just wanted to make sure the same strategy will apply to all the other 249 records coming in the future.

tiborsimko replied:
One more thing: for 2024 MC release, the scripts were creating file list files of this kind:

```
MC_aMcAtNloHerwig7EvtGen_H7UE_SM4topsNLO_filelist.json
MC_aMcAtNloHerwig7EvtGen_MEN30NLO_tHjb125_gamgam_filelist.json
MC_aMcAtNloHerwig7EvtGen_MEN30NLO_ttH125_gamgam_filelist.json
MC_aMcAtNloPy8EG_A14N30NLO_LQd_gstML_0p3_nonallhad_M1000_filelist.json
MC_aMcAtNloPy8EG_A14NNPDF23LO_ppx0yy_FxFx_Np012_SM_filelist.json
MC_aMcAtNloPythia8EvtGen_A14NNPDF23LO_ppx0_FxFx_Np012_SM_filelist.json
```

This gave rise to the file index names such as:

```
root://eospublic.cern.ch//eos/opendata/atlas/rucio/mc20_13TeV/file-indexes/mc20_13TeV_MC_aMcAtNloPy8EG_A14N30NLO_LQd_gstML_0p3_nonallhad_M1000_file_index.json"
```

Now we have:

```
MC_301001_hepmc_filelist.json
MC_301002_hepmc_filelist.json
MC_301003_hepmc_filelist.json
MC_301004_hepmc_filelist.json
MC_301005_hepmc_filelist.json
MC_301006_hepmc_filelist.json
```

This would lead to:

```
root://eospublic.cern.ch//eos/opendata/atlas/rucio/mc16_13TeV/file-indexes/mc16_13TeV_MC_301001_hepmc_file_index.json"
```

Are these names OK? Would you like the file indexes to be named in this way, or are there any better semantically meaningful names coming to these files?

zlmarshall replied:
Thanks for the comments, @tiborsimko ! For your questions:

1) Yeah, I think we need the savings; it's a factor of four because this is a text format

2) Indeed, we'll want file indices, so leaving these for now (but happy to rename them if you like!)

3) Good question. This looks worrying to me (it's a LOT of files), so I've asked the rucio guys to confirm. Meanwhile, you have the naming pattern right, that all the files with the same second number should be grouped. It's possible that there are rare exceptions to this rule, so we should be able to be a tiny bit flexible. (I hope the script gets it all right, though!)

4) Yep, good point. I've made the names considerably longer, but they now have a bunch of useful information in them (some of which is necessary, like the com energy that was previously missing).

Please let me know if I missed anything!

Best,
Zach

zlmarshall replied:
Quick update on item 3: the EOS team apparently said they can handle MANY files in a single directory, so this is the expected and intended behavior. 

I think that's everything, so I'm going to mark this resolved, but please yell if I missed anything!

tiborsimko commented on atlas-2025-odfr-hepmc/make_od_hepmc_json.py:
```diff
@@ -0,0 +1,466 @@
+                                       'type':'isParentOf'
+                                      } ]
+    # Write myself a json file
+    summary_file_name = 'atlas-hepmc-13tev-summary.json'
```
Cosmetics: The generated record JSON file names are:

```
$ ls -1 test/atlas-hepmc-*.json
test/atlas-hepmc-13tev-summary.json
test/atlas-hepmc-specialised-sm-z-boson-13TeV.json
```

i.e. once 13TeV comes as a prefix (for summary record), once as a suffix (for individual datasets).

It would be nicer to use `atlas-hepmc-13tev-....json` file name convention for everything, so that all the files would be grouped together alphabetically by the same prefix, just in case there will be some more HEPMC releases coming later in the future.

Would it be possible to rename individual datasets from `atlas-hepmc-specialised-sm-z-boson-13TeV.json` to `atlas-hepmc-13TeV-specialised-sm-z-boson.json`?

(just a cosmetic comment about the file organisation)

zlmarshall replied:
Yep, I agree, and done now in the latest commit!

Commit:
Comments from Tibor (thanks)

Commit:
Little textual update before the release

zlmarshall commented:
Hi @tiborsimko ,

Thank you for the test deployment! I've pushed a very minor update to the text that would be nice to have in. The only other thing I'd like if possible is for the empty "Files and Indexes" block to be removed from the meta-record at https://opendata-qa.cern.ch/record/160000 . After that, I think you're all clear to press the button and make these public!

Thanks again,
Zach

Commit:
Updating file names to match names on disk

Commit:
Updating description with update notes (ha) and adding a build full md script

Commit:
Suggested changes from Tibor to the description to format it much more nicely

Commit:
Major script update in preparation for major release

Commit:
Updates for scaling up of transfers and samples

Commit:
Working off the file on eos

zlmarshall commented:
Hi @tiborsimko ,

I think this one's now ready for you to take a look at. You can run `make_od_hepmc_json.py` to generate all the files, as usual. I'm using the metadata file stashed on eos now. There are a LOT of json files generated here, and you'll see a familiar structure, I think. There are also quite a few warnings spit out by that script, which will be reduced as we get the remaining files transferred to CERN (there are still a modest number of datasets to go, because one of our grid sites is down for a long stretch). That grid site is expected to be down for so long that my inclination would be to 'just' release this to the portal and then add files once they become available (it would only be adding, no removal and no modification, which I think we agreed was ok).

Please let me know in case you spot an issues, of course!

Thanks,
Zach

zlmarshall commented:
Unfortunately, we've identified a bug affecting a large number of these samples. Please hold off on merging this until we can come back with a plan of attack.

Commit:
Adjusting for eos output file location and extra exotics samples

Commit:
Final updates for release

Samples all transferred, scripts updated and tested; I believe we are
now ready to merge this beast and release the records on the Open Data
Portal!

zlmarshall commented:
Hi @tiborsimko ,

Just a reminder of this one, which has been waiting for a bit more than two months now.

Best,
Zach
